target/linux/generic/files/drivers/net/phy/rtl8367b.c

```diff
@@ -730,39 +516,24 @@ static int rtl8367b_write_phy_reg(struct rtl8366_smi *smi,
 static int rtl8367b_init_regs(struct rtl8366_smi *smi)
 {
 	const struct rtl8367b_initval *initvals;
-	u32 chip_num;
-	u32 chip_ver;
-	u32 rlvid;
 	int count;
-	int err;
-
-	REG_WR(smi, RTL8367B_RTL_MAGIC_ID_REG, RTL8367B_RTL_MAGIC_ID_VAL);
-	REG_RD(smi, RTL8367B_CHIP_NUMBER_REG, &chip_num);
-	REG_RD(smi, RTL8367B_CHIP_VER_REG, &chip_ver);
 
-	if ((chip_ver ==  0x0020 || chip_ver == 0x00A0) && chip_num == 0x6367)  {
+	switch (smi->rtl8367b_chip) {
+	case RTL8367B_CHIP_RTL8363SB:
+	case RTL8367B_CHIP_RTL8367RB:
+	case RTL8367B_CHIP_RTL8367R_VB:
+		initvals = rtl8367b_initvals;
```

Comment:
I think you have to add the check for the `rlvid` here. 
The old code uses `rtl8367c_initvals` for `chip_num == 0x6367` which the new code does in the same way. 
If it is not  `chip_num == 0x6367`  it was checking the `rlvid`. 
If you are not sure that we do not support such chips I would prefer that you add the check here. 

I do not know Realtek product development in detail, but it is common to have multiple chip revisions, the later ones are fixing bugs and need less workarounds in the driver. One chip is often sold under different names for different markets. I assume that all the 3 product names listed here are using the same chip with different fusing and/or packaging and there are 2 revisions which need different initialization.

Reply:
rlvid was tested ror chip_num != 0x6367
It could have three possibilities:
- be equal to 1
- be equal to 0
- other

For 1 the supported chips are: RTL8363SB, RTL8367RB and RTL8367R-VB (chip_ver = 0x**1**000 and 0x**1**010)
For <>0 and 1: an error was generated in rtl8367b_detect (unknow rlvid)
For 0: There are no routers that could belong here.
<body>

 Family RTL8367B | chip | chip_ver | rlvid
-- | -- | -- | --
rt3883_belkin_f9k1109v1 | RTL8367R-VB | 0x1010 | 1
mt7620a_tplink_archer-c2-v1 | RTL8367RB | 0x1000 | 1 
mt7620a_zyxel_keenetic-viva | RTL8367RB | 0x1000 | 1 
mt7620a_iodata_wn-ac733gr3 | RTL8367RB | 0x1000  | 1
rt3662_dlink_dir-645 | RTL8367RB | 0x1000 | 1
Netgear-wndap620 | RTL8363SB | 0x1000 | 1
Netgear-wndap660 | RTL8363SB | 0x1000 | 1
</body>

Reply:
Because I am not sure about RTL8363SB.
I found various information on the Internet, one that has chip_ver=0x6000 and chip_num=0x1000 (https://www.spinics.net/lists/netdev/msg909306.html) and in https://github.com/cgoder/openwrt_rtk/blob/ d237e45ee480ee7e75c80a4c61e1a335a0a77755/rtk_openwrt_sdk/target/linux/rtkmipsel/files/drivers/net/rtl819x/rtl8367r/rtk_api.c#L77C1-L78C5 yet another code.
To be sure not to miss any version of the circuits, I go back to the previous code

Reply:
To be sure I would need to have system logs for Netgear-wndap620 and Netgear-wndap660 for this code

Reply:
I found the log https://forum.openwrt.org/t/wndap660-kernel-panic-during-installation/57770 According to it, the switch is detected as RTL8367RB, so rlvid is equal to 1. So the code without rlvid is correct.

@hauke What do you think, is it better to use a simpler code without rlvid or the code with RTL8367B_CHIP_UNKNOWN_RLVID0 and RTL8367B_CHIP_UNKNOWN_RLVID1 but they are not currently used in any of the routers. Do "spare" types make sense?